### ISSUE TITLE
token_type_ids not getting padded while finetuning llama2

### ISSUE THREAD

**Issue by palash04** [mosaicml/llm-foundry#486]
## Environment

```
OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-144-generic-x86_64-with-glibc2.31
Is CUDA available: N/A
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 470.161.03

transformers: 4.30.2
llm-foundry: 0.2.0
mosaicml: 0.15.1
mosaicml-cli: 0.4.16
mosaicml-streaming: 0.5.1
```
## To reproduce

Yaml config:
```
max_seq_len: 512
global_seed: 17

# Run Name
run_name: # If left blank, will be read from env var $RUN_NAME

# Model
model:
  name: hf_causal_lm
  pretrained_model_name_or_path: meta-llama/Llama-2-7b-hf
  init_device: cpu
  pretrained: true
  use_auth_token: true

# Tokenizer
tokenizer:
  name: meta-llama/Llama-2-7b-hf
  kwargs:
    model_max_length: ${max_seq_len}

# Local data to load into huggingface datasets
dataset: &hf_dataset
  hf_name: json
  hf_kwargs:
    data_dir: /raid/palash.kamble/LLM/finetune_json_data  # /raid/akshat.patidar/json_data
  preprocessing_fn: 

# Dataloaders
train_loader: &train_loader
  name: finetuning
  dataset:
    <<: *hf_dataset
    split: train
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    shuffle: true
    # # Use `python llmfoundry/data/packing.py --yaml-path /path/to/this/yaml/ ...`
    # # to profile this run's optimal packing_ratio as it depends on GPU count,
    # # batch size, sequence length
    # packing_ratio:
  drop_last: true
  num_workers: 8
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

# eval_loader:
#   <<: *train_loader
#   dataset:
#     <<: *hf_dataset
#     split: validation

# Optimization
scheduler:
  name: linear_decay_with_warmup # linear no warmup is HF default which dolly used
  t_warmup: 0ba
  alpha_f: 0

optimizer:
  # mimic HF defaults to replicate dolly
  name: decoupled_adamw
  lr: 1.0e-5
  betas:
  - 0.9
  - 0.999
  eps: 1.0e-8
  weight_decay: 0

algorithms:
  gradient_clipping:
    clipping_type: norm
    clipping_threshold: 1.0

max_duration: 1ep
eval_interval: 1
eval_first: false
eval_subset_num_batches: -1
global_train_batch_size: 128

# System
seed: ${global_seed}
device_eval_batch_size: 4
device_train_microbatch_size: 4
# device_train_microbatch_size: auto
precision: amp_bf16

# FSDP
fsdp_config:
  sharding_strategy: FULL_SHARD
  mixed_precision: PURE
  activation_checkpointing: false
  activation_checkpointing_reentrant: false
  activation_cpu_offload: false
  limit_all_gathers: true
  verbose: false

# Logging
progress_bar: false
log_to_console: true
console_log_interval: 1ba

callbacks:
  speed_monitor:
    window_size: 10
  lr_monitor: {}
  memory_monitor: {}
  runtime_estimator: {}

# loggers:
#   wandb: {}

# Checkpoint to local filesystem or remote object store
save_interval: 2ba
save_num_checkpoints_to_keep: 1  # Important, this cleans up checkpoints saved to DISK
save_folder: /raid/palash.kamble/LLM/finetune/checkpoints/llama2_7b

```

## Run command
```
TRAIN_PATH="/raid/palash.kamble/LLM/llm-foundry/scripts/train/train.py"
YAML_PATH="/raid/palash.kamble/LLM/llm-foundry/scripts/train/yamls/finetune/llama2-7b_local_data_sft.yaml"
python3 -m composer -n 8 $TRAIN_PATH $YAML_PATH
```

## Error
```
ValueError: Caught ValueError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/raid/palash.kamble/miniconda3/envs/llm_foundry/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 744, in convert_to_tensors
    tensor = as_tensor(value)
ValueError: expected sequence of length 481 at dim 1 (got 477)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/raid/palash.kamble/miniconda3/envs/llm_foundry/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/raid/palash.kamble/miniconda3/envs/llm_foundry/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 54, in fetch
    return self.collate_fn(data)
  File "/raid/palash.kamble/LLM/llm-foundry/llmfoundry/data/finetuning/collator.py", line 116, in __call__
    batch = self._process_and_batch_decoder_only(examples)
  File "/raid/palash.kamble/LLM/llm-foundry/llmfoundry/data/finetuning/collator.py", line 222, in _process_and_batch_decoder_only
    batch = self.tokenizer.pad(
  File "/raid/palash.kamble/miniconda3/envs/llm_foundry/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 3071, in pad
    return BatchEncoding(batch_outputs, tensor_type=return_tensors)
  File "/raid/palash.kamble/miniconda3/envs/llm_foundry/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 211, in __init__
    self.convert_to_tensors(tensor_type=tensor_type, prepend_batch_axis=prepend_batch_axis)
  File "/raid/palash.kamble/miniconda3/envs/llm_foundry/lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 760, in convert_to_tensors
    raise ValueError(
ValueError: Unable to create tensor, you should probably activate truncation and/or padding with 'padding=True' 'truncation=True' to have batched tensors with the same length. Perhaps your features (`token_type_ids` in this case) have excessive nesting 
(inputs type `list` where type `int` is expected).
```

Upon checking **input_ids**, **labels**, **attention_mask**, these are of length equal to **seq_len**, but only **token_type_ids** is having variable length.

## I did the following temporarily to resolve the issue but don't think it is the right way.
Go to /miniconda3/envs/llm_foundry/lib/python3.10/site-packages/transformers/tokenization_utils_base.py
Added following lines to function **convert_to_tensors**
```
seq_len = 512 # taken from yaml
# pad token_type_ids
for k, v in self.items():
    if k == 'token_type_ids':
        for idd in range(len(v)):
            self[k][idd] = [0] * seq_len
```

I think the custom collate function **Seq2SeqFinetuningCollator** should take care of padding **token_type_ids**.

Or is there a way to mention 'padding=True' 'truncation=True'?

**Comment by dakinggg** [comment]
The thing that I am confused about is why there are `token_type_ids` at all. The llama tokenizer doesn't use `token_type_ids` as far as I know. Do you know where the `token_type_ids` are coming from?

**Comment by dakinggg** [comment]
I tracked this down to a transformers but that was fixed in `4.31.0` (https://github.com/huggingface/transformers/pull/24042/files). If you manually update transformers with `pip install --upgrade transformers`, it should work. We are also updating our transformers pins, but for now you can workaround with the manual install. Thanks for reporting!

**Comment by palash04** [comment]
Thanks @dakinggg.
